### ISSUE TITLE
Rancher 2.2 - Enabling Prometheus causes GUI crash

### ISSUE THREAD

**Issue by cjellick** [rancher/rancher#19274]
see https://forums.rancher.com/t/rancher-2-2-enabling-prometheus-causes-gui-crash/13757/6

**Comment by loganhz** [comment]
We cannot reproduce this issue. Rancher didn't crash after enabling monitoring

**Comment by jiaqiluo** [comment]
I cannot reproduce the issue on v2.2.0
Steps:
- run Rancher v2.2.0
- add an EKS cluster with all default settings
- enable the cluster-level monitoring

Result:
- the cluster-level monitoring works as expected 
- docker stat is normal
```
root@jiaqi-220:~# sudo docker stats -a --no-stream
CONTAINER           CPU %               MEM USAGE / LIMIT       MEM %               NET I/O             BLOCK I/O           PIDS
5abb71e80ae5        27.13%              506.7 MiB / 7.796 GiB   6.35%               11.4 MB / 10.8 MB   0 B / 69.4 MB       29
root@jiaqi-220:~# sudo docker stats -a --no-stream
CONTAINER           CPU %               MEM USAGE / LIMIT       MEM %               NET I/O             BLOCK I/O           PIDS
5abb71e80ae5        7.09%               507.4 MiB / 7.796 GiB   6.36%               11.9 MB / 11.2 MB   0 B / 73.9 MB       29
root@jiaqi-220:~# sudo docker stats -a --no-stream
CONTAINER           CPU %               MEM USAGE / LIMIT       MEM %               NET I/O             BLOCK I/O           PIDS
5abb71e80ae5        5.32%               507.5 MiB / 7.796 GiB   6.36%               12.1 MB / 11.4 MB   0 B / 75.7 MB       29
root@jiaqi-220:~# sudo docker stats -a --no-stream
CONTAINER           CPU %               MEM USAGE / LIMIT       MEM %               NET I/O             BLOCK I/O           PIDS
5abb71e80ae5        6.34%               507.5 MiB / 7.796 GiB   6.36%               12.1 MB / 11.4 MB   0 B / 75.8 MB       29
```

**Comment by alena1108** [comment]
@loganhz can you follow up with the user on forums to provide Rancher server logs as the crash was observed on the rancher management side.

**Comment by jkuspa** [comment]
@loganhz Thanks for trying this.  Here are the steps I took. 
Let me know if I can get you anything else. 

Setup: 
`docker run -d --restart=unless-stopped --name rancher --hostname rancher -p 80:80 -p 443:443 rancher/rancher:latest`

- Validate server version: 
https://localhost/v3/settings/server-version

```
"name": "server-version",
"source": "env",
"type": "setting",
"uuid": "bba6a878-5563-11e9-9b7b-0242ac110002",
"value": "v2.2.0"
```

Steps:

- add an EKS cluster with all default settings
- enable the cluster-level monitoring

Result: 

- Docker Stats:  
```
CONTAINER ID        NAME                CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
c9174534e721        rancher             241.25%             614.5MiB / 3.855GiB   15.57%              24.8MB / 15.3MB     97.3MB / 152MB      76
```

- Logs: 
`docker logs -f containerID`  

```
[main] 2019/04/02 16:53:05 Starting Tiller v2.10+unreleased (tls=false)
[main] 2019/04/02 16:53:05 GRPC listening on :33720
[main] 2019/04/02 16:53:05 Probes listening on :40986
[main] 2019/04/02 16:53:05 Storage driver is ConfigMap
[main] 2019/04/02 16:53:05 Max history per release is 0
[tiller] 2019/04/02 16:53:05 getting history for release monitoring-operator
[storage] 2019/04/02 16:53:05 getting release history for "monitoring-operator"
[storage/driver] 2019/04/02 16:53:05 query: failed to query with labels: the server is currently unable to handle the request (get configmaps)
[tiller] 2019/04/02 16:53:05 preparing update for monitoring-operator
[storage] 2019/04/02 16:53:05 getting deployed releases from "monitoring-operator" history
[storage/driver] 2019/04/02 16:53:05 query: failed to query with labels: the server is currently unable to handle the request (get configmaps)
2019/04/02 16:53:05 [ERROR] AppController p-8lnj9/monitoring-operator [helm-controller] failed with : failed to install app monitoring-operator. Error: UPGRADE FAILED: the server is currently unable to handle the request (get configmaps)

[main] 2019/04/02 16:53:05 Starting Tiller v2.10+unreleased (tls=false)
[main] 2019/04/02 16:53:05 GRPC listening on :35841
[main] 2019/04/02 16:53:05 Probes listening on :36811
[main] 2019/04/02 16:53:05 Storage driver is ConfigMap
[main] 2019/04/02 16:53:05 Max history per release is 0
[tiller] 2019/04/02 16:53:06 getting history for release cluster-monitoring
[storage] 2019/04/02 16:53:06 getting release history for "cluster-monitoring"
[storage/driver] 2019/04/02 16:53:06 query: failed to query with labels: the server is currently unable to handle the request (get configmaps)
[tiller] 2019/04/02 16:53:06 preparing update for cluster-monitoring
[storage] 2019/04/02 16:53:06 getting deployed releases from "cluster-monitoring" history
[storage/driver] 2019/04/02 16:53:06 query: failed to query with labels: the server is currently unable to handle the request (get configmaps)
2019/04/02 16:53:06 [ERROR] AppController p-8lnj9/cluster-monitoring [helm-controller] failed with : failed to install app cluster-monitoring. Error: UPGRADE FAILED: the server is currently unable to handle the request (get configmaps)

[tiller] 2019/04/02 16:53:06 getting history for release monitoring-operator
[storage] 2019/04/02 16:53:06 getting release history for "monitoring-operator"
[storage/driver] 2019/04/02 16:53:06 query: failed to query with labels: the server is currently unable to handle the request (get configmaps)
[tiller] 2019/04/02 16:53:06 preparing update for monitoring-operator
[storage] 2019/04/02 16:53:06 getting deployed releases from "monitoring-operator" history
[storage/driver] 2019/04/02 16:53:06 query: failed to query with labels: the server is currently unable to handle the request (get configmaps)
2019/04/02 16:53:06 [ERROR] AppController p-8lnj9/monitoring-operator [helm-controller] failed with : failed to install app monitoring-operator. Error: UPGRADE FAILED: the server is currently unable to handle the request (get configmaps)
```

- `ls` of /tmp, helm continuously tries to redeploy
```
root@rancher:/tmp# date && ls -arlt
Tue Apr  2 17:03:55 UTC 2019
total 36
drwxr-xr-x 1 root root 4096 Apr  2 16:23 ..
drwx------ 3 root root 4096 Apr  2 17:01 helm-294567211
drwx------ 3 root root 4096 Apr  2 17:01 kubeconfig-168774896
drwx------ 3 root root 4096 Apr  2 17:01 helm-279872399
srwxr-xr-x 1 root root    0 Apr  2 17:01 log.sock
drwx------ 3 root root 4096 Apr  2 17:03 helm-332543872
drwx------ 3 root root 4096 Apr  2 17:03 kubeconfig-641771743
drwx------ 3 root root 4096 Apr  2 17:03 helm-067094706
drwx------ 3 root root 4096 Apr  2 17:03 kubeconfig-003345513
drwxrwxrwt 1 root root 4096 Apr  2 17:03 .
root@rancher:/tmp# date && ls -arlt
Tue Apr  2 17:03:56 UTC 2019
total 36
drwxr-xr-x 1 root root 4096 Apr  2 16:23 ..
drwx------ 3 root root 4096 Apr  2 17:01 helm-294567211
drwx------ 3 root root 4096 Apr  2 17:01 kubeconfig-168774896
drwx------ 3 root root 4096 Apr  2 17:01 helm-279872399
srwxr-xr-x 1 root root    0 Apr  2 17:01 log.sock
drwx------ 3 root root 4096 Apr  2 17:03 helm-965848244
drwx------ 3 root root 4096 Apr  2 17:03 kubeconfig-288911491
drwx------ 3 root root 4096 Apr  2 17:03 helm-925957416
drwx------ 3 root root 4096 Apr  2 17:03 kubeconfig-576864871
drwxrwxrwt 1 root root 4096 Apr  2 17:03 .
```


Rancher GUI after enabling monitoring. 

![Screen Shot 2019-04-02 at 1 07 21 PM](https://user-images.githubusercontent.com/15388693/55421903-6448a700-5548-11e9-865b-10be693c121a.png)

**Comment by jkuspa** [comment]
Not sure if this is related or not...
https://github.com/rancher/rancher/issues/16039#issuecomment-435501062

**Comment by loganhz** [comment]
Hi @jkuspa could you provide more info, please?

when you do the `docker stats`, the CPU is always very high? 
Can you try to deploy another app in Library catalog please?
How long does it take from you enable monitoring and Rancher is down?
What's the size of node in EKS and how many nodes in the EKS cluster?
Where is your rancher server running?

**Comment by jkuspa** [comment]
Docker stats are only high once monitoring is enabled, here they are after the cluster is torn down. 
```
CONTAINER ID        NAME                CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
c9174534e721        rancher             29.02%              445.4MiB / 3.855GiB   11.28%              8.92MB / 6.86MB     295kB / 222MB       30
```

Time from enabling monitoring to Rancher going down: Roughly 20 minutes. 
Node Size: m4.large?
Node Count: 3
AWS Region: us-east-1
Rancher server: Running locally on a Macbook Pro
```
docker version
Client: Docker Engine - Community
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        6247962
 Built:             Sun Feb 10 04:12:39 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       6247962
  Built:            Sun Feb 10 04:13:06 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

Deploy another app: 
I'll try to do this later and post the results.

**Comment by loganhz** [comment]
We didn't test rancher server on macbook. 
Could you try to run rancher server on ubuntu, please?
Thanks.

**Comment by loganhz** [comment]
It looks like the issue is `[storage/driver] 2019/04/02 16:53:06 query: failed to query with labels: the server is currently unable to handle the request (get configmaps)`

So we need to find out why your cluster is in an unavailable state first.

**Comment by jkuspa** [comment]
@loganhz 
Correction on where Rancher is running:
Rancher is running via Docker, which is using Ubuntu.
```
cat /etc/os-release 
NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
```

I'll try the following and update this ticket. 
- Create a rancher cluster in an AWS account with more permissions
- Tail all logs to a file to capture the errors when monitoring is enabled.

**Comment by alena1108** [comment]
https://github.com/rancher/rancher/pull/19079 can mitigate the issue

**Comment by jkuspa** [comment]
@loganhz @alena1108 


To troubleshoot this, I performed the following steps. 

1. Reinstalled Rancher via the above docker commands. 

2. Started a new cluster in an AWS account with a user that had full administrative rights. 

3. Installed an app (Redis), as requested, via the Rancher GUI. 
Global > Apps > Enabled Helm charts > Installed Redis

![Screen Shot 2019-04-04 at 10 28 39 AM](https://user-images.githubusercontent.com/15388693/55563918-cd552980-56c4-11e9-8933-8d43b6b42049.png)

4. Tailed logs to a file and enabled Monitoring
If you want full logs I can upload the file. 

`docker logs -f containerID >> rancher.log`


Oddly enough, the first ERROR below matches this issue, but this is NOT an airgap installation: https://github.com/rancher/rancher/issues/19362

```
2019/04/04 14:20:12 [INFO] Creating roleBinding User user-8hfzj Role project-owner
2019/04/04 14:20:12 [INFO] Creating roleBinding User user-8hfzj Role admin
2019/04/04 14:20:12 [INFO] Creating roleBinding User u-n7qtawqjbi Role project-member
2019/04/04 14:20:12 [INFO] Creating roleBinding User u-n7qtawqjbi Role edit
2019/04/04 14:20:12 [INFO] Creating token for user user-8hfzj
Release "cluster-monitoring" does not exist. Installing it now.
Release "monitoring-operator" does not exist. Installing it now.
2019/04/04 14:20:15 [ERROR] AppController p-nkslk/cluster-monitoring [helm-controller] failed with : failed to install app cluster-monitoring. Error: validation failed: [unable to recognize "": no matches for kind "Prometheus" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"]

2019/04/04 14:20:15 [ERROR] AppController p-nkslk/monitoring-operator [helm-controller] failed with : failed to install app monitoring-operator. Error: validation failed: unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"

Release "cluster-monitoring" does not exist. Installing it now.
Release "monitoring-operator" does not exist. Installing it now.
2019/04/04 14:20:17 [INFO] kontainerdriver amazonelasticcontainerservice listening on address 127.0.0.1:10476
2019/04/04 14:20:17 [INFO] kontainerdriver amazonelasticcontainerservice stopped
2019/04/04 14:20:17 [ERROR] ClusterController c-snx5k [temporary-credentials] failed with : error updating temporary credentials annotation: Operation cannot be fulfilled on clusters.management.cattle.io "c-snx5k": the object has been modified; please apply your changes to the latest version and try again
2019/04/04 14:20:17 [ERROR] AppController p-nkslk/monitoring-operator [helm-controller] failed with : failed to install app monitoring-operator. Error: validation failed: unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"

2019/04/04 14:20:17 [ERROR] AppController p-nkslk/cluster-monitoring [helm-controller] failed with : failed to install app cluster-monitoring. Error: validation failed: [unable to recognize "": no matches for kind "Prometheus" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"]

Release "monitoring-operator" does not exist. Installing it now.
Release "cluster-monitoring" does not exist. Installing it now.
2019/04/04 14:20:18 [ERROR] ClusterController c-snx5k [cluster-monitoring-handler] failed with : unable to sync Cluster c-snx5k(rancher-clarity): failed to detect the installation status of monitoring components: [Grafana Deployment isn't deployed, Prometheus StatefulSet isn't deployed, Kube State Exporter Deployment isn't deployed, Node Exporter DaemonSet isn't deployed]
2019/04/04 14:20:19 [ERROR] AppController p-nkslk/monitoring-operator [helm-controller] failed with : failed to install app monitoring-operator. Error: validation failed: unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"

Release "monitoring-operator" does not exist. Installing it now.
2019/04/04 14:20:19 [ERROR] AppController p-nkslk/cluster-monitoring [helm-controller] failed with : failed to install app cluster-monitoring. Error: validation failed: [unable to recognize "": no matches for kind "Prometheus" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"]

```

**Comment by deniseschannon** [comment]
@tfiduccia can you help to try and reproduce @jkuspa steps?

**Comment by deniseschannon** [comment]
@jkuspa we haven't been able to reproduce but we've put some safeguards to prevent tiller from getting in a crash loop. would you be willing to try to see if you can get it working with v2.2.3-rc6?

**Comment by dsargent3220** [comment]
I don't know if this helps or not, but I ran into this same error when i upgraded my HA Rancher test cluster for 2.2.8 to 2.3.1 but I think it was because I flubbed the config the first time.  I set the hostname to an incorrect URL and immediately noticed it, so I then ran the upgrade again with the correct hostname.  Rancher worked, but when I tried to click o a cluster I got the above error.  I waited for the upgrade to complete and everything to settle down but always got the monitoring error.  Funny thing is, these are really new clusters, I've never installed monitoring on them.
I used helm to rollback to the previous 2.2.8 and did the upgrade again (correctly) and everything worked.  Might not be related at all, but thought I should share.
